### ISSUE TITLE
"sqlfluff fix" fails with "Some operations failed... Some errors cannot be fixed or there is another error blocking it."

### ISSUE THREAD

**Issue by barrywhart** [sqlfluff/sqlfluff#1296]
-- If this is a parsing or linting issue, please include a minimal SQL example which reproduces the issue, along with the `sqlfluff parse` output, `sqlfluff lint` output and `sqlfluff fix` output when relevant. Finally, please add any labels that are related to the issue before submitting so that it can be addressed easily! --

## Expected Behaviour

File successfully fixed (or left untouched).

## Observed Behaviour

`sqlfluff fix` reports an error:
```
 sqlfluff fix sql/campaign_performance_queries/constraints/_postcondition_ignore_internal_accounts.sql
==== finding fixable violations ====
== [sql/campaign_performance_queries/constraints/_postcondition_ignore_internal_accounts.sql] FAIL
L:   9 | P:  59 | L016 | Line is too long
==== fixing violations ====
1 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] ...
Attempting fixes...
Persisting Changes...
== [sql/campaign_performance_queries/constraints/_postcondition_ignore_internal_accounts.sql] FAIL
Done. Some operations failed. Please check your files to confirm.
Some errors cannot be fixed or there is another error blocking it.
```

The exit code reports failure, which means that using `sqlfluff fix` in a pre-commit hook prevents a user from committing the file in this case. Two suggestions:
* Fix the underlying bug
* Consider whether `sqlfluff fix` should instead exit with status `0` in this case, because it's not clear the user can do anything about this.

## Steps to Reproduce

```
-- Checks to make sure that no internal accounts are included in the campaign_performance table.
SELECT user_id
FROM
    `{{gcp_project}}.{{dataset}}.campaign_performance`
WHERE
    user_id IN (
        SELECT user_id
        FROM
            `{{BI_INTERNAL_USERS}}` {{table_at_job_start}}
    )
```

## Version
SQLFluff: 0.6.3
Python: 3.7.9

## Configuration
```
[sqlfluff]
verbose = 0
nocolor = False
dialect = bigquery
templater = jinja
rules = None
exclude_rules = L019,L025,L027,L029,L031,L032,L034
recurse = 0

[sqlfluff:templater:jinja:macros]

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 120

# Some rules have their own specific config.
[sqlfluff:rules:L010]
capitalisation_policy = consistent

[sqlfluff:rules:L016]
max_line_length = 100

[sqlfluff:rules:L014]
capitalisation_policy = consistent

[sqlfluff:rules:L030]
capitalisation_policy = consistent

[sqlfluff:rules:L047]
prefer_count_1 = True

# Settings in this section will be provided to SQL templates during the SQL
# formatting/linting process. Generally, the exact values don't matter as long
# as they result in valid SQL syntax when substituted in the file.
[sqlfluff:templater:jinja:context]

# DEFAULT VALUES: Overridden in script/run_job.sh.
# Best practice: Use lower_case for variables
gcp_project=gcp_project
dataset=dataset
table_at_job_start=FOR SYSTEM_TIME AS OF CAST('2020-05-11T14:02:52+00:00' AS TIMESTAMP)

BI_INTERNAL_USERS=mc-business-intelligence.bi_reporting.internal_users
```

**Comment by barrywhart** [comment]
May be related to #848 (same error message).

**Comment by barrywhart** [comment]
I think this "failure" occurs because of an "unfixable" issue blocking the fix from being applied. See this output from `sqlfluff lint` on the same file:
```
$ sqlfluff lint sql/campaign_performance_queries/constraints/_postcondition_ignore_internal_accounts.sql
== [sql/campaign_performance_queries/constraints/_postcondition_ignore_internal_accounts.sql] FAIL
L:   9 | P:  37 | L046 | Jinja tags should have a single whitespace on either
                       | side.
L:   9 | P:  59 | L016 | Line is too long
```

This looks incorrect: The Jinja tag being referenced _does_ have a whitespace on the left side, and the right side ends with a newline, so a whitespace seems unnecessary there.

**Comment by barrywhart** [comment]
Even if I add `L046` to the `exclude_rules` list, I still get the same error. IIRC, `sqlfluff fix` has some limitations about changing code near Jinja tags. That may be the reason for the failure, in which case, the correctness (or potential incorrectness) of`L046` is probably not relevant to addressing _this_ issue.

**Comment by tunetheweb** [comment]
> This looks incorrect: The Jinja tag being referenced _does_ have a whitespace on the left side, and the right side ends with a newline, so a whitespace seems unnecessary there.

I don't agree with you here. Jinja tags should like `{{ this }}` and not `{{this}`, so your code should be this to pass rule L046:

```sql
-- Checks to make sure that no internal accounts are included in the campaign_performance table.
SELECT user_id
FROM
    `{{ gcp_project }}.{{ dataset }}.campaign_performance`
WHERE
    user_id IN (
        SELECT user_id
        FROM
            `{{ BI_INTERNAL_USERS }}` {{ table_at_job_start }}
    )
```

When you change it to that it complains of just the line length error, though `sqlfluff fix` still throws the failure.

To me the error is because line 9 becomes this, when the Jinja tags are substituted:

```
            `mc-business-intelligence.bi_reporting.internal_users` FOR SYSTEM_TIME AS OF CAST('2020-05-11T14:02:52+00:00' AS TIMESTAMP)
```

This line is indeed more than the 100 character `max_line_length` in your .sqlfluff file, but how is sqlfluff supposed to fix this? It only sees the error **AFTER** variable substitution at which point it would struggle to map this back to the original SQL to apply the fix.

Manually changing the code to this avoids the error:

```sql
-- Checks to make sure that no internal accounts are included in the campaign_performance table.
SELECT user_id
FROM
    `{{ gcp_project }}.{{ dataset }}.campaign_performance`
WHERE
    user_id IN (
        SELECT user_id
        FROM
            `{{ BI_INTERNAL_USERS }}`
            {{ table_at_job_start }}
    )
```

I'm not surprised that sqlfluff struggles with this one (and other Jinja templating issues to be honest) and not sure it ever will be able to cope. It could ignore unfixable errors like this I suppose but L016 **IS** supposed to be autofixable, but in this case it isn't - despite there clearly being a violation. So it gives up.

**Comment by barrywhart** [comment]
Oh, I thought it wanted spaces around the outside of the curly braces, not the inside. I'm not sure I like this rule when applied to simple variable substitutions (not a complex expression), but that's a matter of taste, and we can just disable that rule if we like. Thanks for the insight!

Any thoughts on the larger concern I raised, about whether "sqlfluff fix" should report a failure exit code when there's an unfixable problem? I suppose if the problem can be manually fixed, it's okay. I'm trying to think about this from the perspective of a SQL developer who's not an expert at either SQL or Sqlfluff -- I want the pre-commit check to be a helper, not a potentially confusing blocker. (For context, I'm an engineer on a 6-person team that supports a team of about 16 data scientists. If they're confused, we help them, but that means we have distractions and get less of our long-term work done.

**Comment by barrywhart** [comment]
I wonder why L046 only warns at one place. In my example query, there are 4 sets of Jinja tags, and none of them have spaces. 🤔

**Comment by tunetheweb** [comment]
As I say I think this is a tough one as L016 is one that should be autofixable, but it gets complicated in cases like this when you introduce Jinja templating.

I guess it could be configured to pass without error, but then you might get others complaining it's not fixing a rule that's supposed to be fixable.

To me this is a bad configuration issue. You've specified a (reasonable) `max_line_length` but then put in really long jinja templating values that result in breaking that. I would put shorter ones in. As the comment in the .sqlfluff file rightly says "Generally, the exact values don't matter as long as they result in valid SQL syntax when substituted in the file." so why bring this pain upon yourself by using really long ones?

Now whether that's scalable to others (you're a maintainer of this project and couldn't figure this out - so what chance does a normal user have) is a good question, and maybe something should be done here. Or the error message improved? though on re-reading it after figuring this out, it does say what the problem is.

> I wonder why L046 only warns at one place. In the example query, there are 4 sets of Jinja tags, and none of them have spaces. 🤔

That is a good question!!

**Comment by barrywhart** [comment]
I think the line length check should applied *before* templating is applied. The line length limit is intended to make the SQL more readable for humans, who presumably are reading the source code, pre-expansion. I created an issue for this a few months ago: #845.

**Comment by tunetheweb** [comment]
That seems fair - and would avoid this issue.

Shorter term though, I'd just user shorter dummy variables. 😀

**Comment by barrywhart** [comment]
I have used shorter variables in the past. We're starting to look at a different workflow, where the `.sqlfluff` file does "double duty", driving not only SQLFluff but also the actual code generation. Developers can override values from `.sqlfluff` if they need to, but I'd like to reserve that for things that vary at runtime, e.g things like the current date.

TBH, I'd like to switch our projects to using dbt rather than our custom tool set that uses Jinja plus some internal tools, but that probably won't happen until 2022 at least.
